test/core/expectations/compound_data/test_expectations_cursor.pkb

```diff
@@ -1041,5 +1041,516 @@ Rows: [ 2 differences ]%
+    open l_expected for select rownum as rni, 'x' SOME_COL  from dual a connect by level < 4;
+     
+    --Act
+    ut3.ut.expect(l_actual).to_equal(l_expected).join_by('RNI');
+    --Assert
+ l_expected_message := q'[%Actual: refcursor [ count = 3 ] was expected to equal: refcursor [ count = 3 ]%
+Diff:%
+%Unable to join sets:%
+%Join key /*/RNI does not exists in actual%]';
+    l_actual_message := ut3.ut_expectation_processor.get_failed_expectations()(1).message;
+    --Assert
+    ut.expect(l_actual_message).to_be_like(l_expected_message);
+  end;
+ 
+  procedure cursor_joinby_compare_1000 is
+    l_actual   SYS_REFCURSOR;
+    l_expected SYS_REFCURSOR;
+  begin
+    --Arrange
+    open l_actual for select object_name from all_objects where rownum <=1100;
+    open l_expected for select object_name from all_objects where rownum <=1100;
+    --Act
+    ut3.ut.expect(l_actual).to_equal(l_expected).join_by('OBJECT_NAME');
+    --Assert
+    ut.expect(expectations.failed_expectations_data()).to_be_empty();
+  end;
+ 
+  procedure cursor_joinby_compare_fail is
+    l_actual   SYS_REFCURSOR;
+    l_expected SYS_REFCURSOR;
+    l_actual_message   varchar2(32767);
+    l_expected_message varchar2(32767);
+  begin
+    --Arrange
+    open l_expected for select username, user_id from all_users union all
+    select 'TEST' username, -600 user_id from dual order by 1 desc;
+    
+    open l_actual for select username, user_id from all_users union all
+    select 'TEST' username, -610 user_id from dual order by 1 asc;
+     
+    --Act
+    ut3.ut.expect(l_actual).to_equal(l_expected).join_by('USERNAME');
+    --Assert
+ l_expected_message := q'[%Actual: refcursor [ count = % ] was expected to equal: refcursor [ count = % ]
+%Diff:%
+%Rows: [ 1 differences ]%
+%Expected: <USER_ID>-600</USER_ID> for key: TEST%
+%Actual:   <USER_ID>-610</USER_ID> for key: TEST%]';
+    l_actual_message := ut3.ut_expectation_processor.get_failed_expectations()(1).message;
+    --Assert
+    ut.expect(l_actual_message).to_be_like(l_expected_message);
+  end;
+  
+  procedure cursor_joinby_cmp_twocol_fail is
+    l_actual   SYS_REFCURSOR;
+    l_expected SYS_REFCURSOR;
+    l_actual_message   varchar2(32767);
+    l_expected_message varchar2(32767);
+  begin
+    --Arrange
+    open l_expected for select username, user_id from all_users union all
+    select 'TEST' username, -600 user_id from dual order by 1 desc;
+    
+    open l_actual for select username, user_id from all_users union all
+    select 'TEST' username, -610 user_id from dual order by 1 asc;
+     
+    --Act
+    ut3.ut.expect(l_actual).to_equal(l_expected).join_by(ut3.ut_varchar2_list('USERNAME,USER_ID'));
+    --Assert
+ l_expected_message := q'[%Actual: refcursor [ count = % ] was expected to equal: refcursor [ count = % ]
+%Diff:%
+%Rows: [ 2 differences ]%
+%Missing:  <USERNAME>TEST</USERNAME><USER_ID>-600</USER_ID> for key: -600:TEST%
```

Comment:
I have a doubt about readability of this message to the user.
` Missing:  <USERNAME>TEST</USERNAME><USER_ID>-600</USER_ID> for key: -600:TEST`
If we don't join by PK, we use syntax like
`  Row No. 3 - Missing:  <RN>3</RN>`
How about if we would present it to the user in a similar way:
`PK <USERNAME>TEST</USERNAME><USER_ID>-600</USER_ID> missing.`

Reply:
I think we should exclude PK columns from the Missing/Extra data list - this is duplicated info.

Reply:
Agree, will update

Reply:
Make sense